src/StarcounterClientFiles/wwwroot/sys/Palindrom/src/palindrom.js

```diff
@@ -34,951 +33,1014 @@ const NodeWebSocket = require('websocket').w3cwebsocket;
-        this._ws.send(msg);
-        this.onSend(msg, this._ws.url, 'WS');
-      } else {
-        const url = this.remoteUrl.href;
-        this.xhr(url, 'application/json-patch+json', msg, (res, method) => {
-          this.onReceive(res.data, url, method);
+        // TODO:(tomalec)[cleanup] hide from public API.
+        setRemoteUrl(remoteUrl) {
+            if (
+                this.remoteUrlSet &&
+                this.remoteUrl &&
+                this.remoteUrl != remoteUrl
+            ) {
+                const message = [
+                    'Session lost.',
+                    'Server replied with a different session ID than the already set one.',
+                    'Possibly a server restart happened while you were working.',
+                    'Please reload the page.',
+                    'Previous session ID: ' + this.remoteUrl,
+                    'New session ID: ' + remoteUrl
+                ].join('\n');
+
+                throw new PalindromError(message);
+            }
+            this.remoteUrlSet = true;
+            this.remoteUrl = new URL(remoteUrl, this.remoteUrl.href);
+        }
+
+        handleSuccessResponse(res, callback) {
+            /* Axios always returns lowercase headers */
+            const location =
+                res.headers &&
+                (res.headers['x-location'] || res.headers['location']);
+            if (location) {
+                this.setRemoteUrl(location);
+            }
+            callback && callback.call(this, res, res.config.method.toUpperCase());
+        }
+
+        /**
+         * Internal method to perform XMLHttpRequest
+         * @param url (Optional) URL to send the request. If empty string, undefined or null given - the request will be sent to window location
+         * @param accept (Optional) HTTP accept header
+         * @param data (Optional) Data payload
+         * @param [callback(response)] callback to be called in context of palindrom with response as argument
+         * @returns {XMLHttpRequest} performed XHR
+         */
+        xhr(url, accept, data, callback, setReferer) {
+            const method = data ? 'PATCH' : 'GET';
+            const headers = {};
+            let requestPromise;
+
+            if (data) {
+                headers['Content-Type'] = 'application/json-patch+json';
+            }
+            if (accept) {
+                headers['Accept'] = accept;
+            }
+            if (this.remoteUrl && setReferer) {
+                headers['X-Referer'] = this.remoteUrl.pathname;
+            }
+            if (method === 'GET') {
+                requestPromise = axios.get(url, {
+                    headers
+                });
+            } else {
+                requestPromise = axios.patch(url, data, {
+                    headers
+                });
+            }
+            requestPromise
+                .then(res => {
+                    this.handleSuccessResponse(res, callback);
```

Comment:
What is the reason you moved callback to function argument? It gives an impression that it will be called asynchrounously or conditionally, but it's not.